### PR TITLE
Support to verify license

### PR DESCRIPTION
**Pull request by Linary**
Change-Id: Ia3fb6b38a032442e1d8bf7cee4d7b4032dc219df

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/license/MachineInfo.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+    public List<String> getIpAddress() {
+        List<String> result = null;
+        List<InetAddress> inetAddresses = this.getLocalAllInetAddress();
+        if (inetAddresses != null && inetAddresses.size() > 0) {
+            result = inetAddresses.stream().map(InetAddress::getHostAddress)
+                                  .distinct().map(String::toLowerCase)
+                                  .collect(Collectors.toList());
+        }
+        return result;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/license/MachineInfo.java]: don't return null

File: hugegraph-core/src/main/java/com/baidu/hugegraph/license/MachineInfo.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            Set<String> uniqueValues = new HashSet<>();
+            for (InetAddress inetAddress : inetAddresses) {
+                String macByInetAddress = this.getMacByInetAddress(inetAddress);
+                if (uniqueValues.add(macByInetAddress)) {
+                    list.add(macByInetAddress);
+                }
+            }
+            result = list;
+        }
+        return result;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/license/MachineInfo.java]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/license/MachineInfo.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    public List<String> getIpAddress() {
+        List<String> result = null;
+        List<InetAddress> inetAddresses = this.getLocalAllInetAddress();
+        if (inetAddresses != null && inetAddresses.size() > 0) {
+            result = inetAddresses.stream().map(InetAddress::getHostAddress)
+                                  .distinct().map(String::toLowerCase)
+                                  .collect(Collectors.toList());
+        }
+        return result;
+    }
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/license/MachineInfo.java]: cache the results for a while

File: hugegraph-core/src/main/java/com/baidu/hugegraph/license/CustomCheckParam.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ * under the License.
+ */
+
+package com.baidu.hugegraph.license;
+
+import java.util.List;
+
+public class CustomCheckParam {
+
+
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/license/CustomCheckParam.java]: add cpu and memory limit

File: hugegraph-api/src/main/java/com/baidu/hugegraph/server/ApplicationConfig.java [hunk 5]
```diff
@@ -67,6 +67,10 @@ public ApplicationConfig(HugeConfig conf) {
         register(new InstrumentedResourceMethodApplicationListener(registry));
     }
 
+    private void installLicense() {
+//        LicenseVerifier.instance().install();
+    }
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/server/ApplicationConfig.java]: remove unused method

**Comment by javeme** [comment]
1 set key: server-1, server-2, ..., hubble-1, hubble-2, ...
2 limit cpu/ram
3 split code and move to common

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/filter/LoadDetectFilter.java [hunk 6]
```diff
@@ -56,14 +58,22 @@
...
 
     @Override
     public void filter(ContainerRequestContext context) {
-        LicenseVerifier.instance().verify();
+        Instant now = Instant.now();
+        Duration interval = Duration.between(lastCheckTime, now);
+        if (!interval.minus(CHECK_INTERVAL).isNegative()) {
+            LicenseVerifier.instance(null).verify();
+            lastCheckTime = now;
+        }
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/filter/LoadDetectFilter.java]: add method LicenseVerifier.verifyIfNeeded()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/util/JsonUtil.java [hunk 7]
```diff
@@ -116,6 +117,18 @@ public static String toJson(Object object) {
...
     }
 
+    public static <T> T  fromJson(InputStream stream, Class<T> clazz) {
+        E.checkState(stream != null,
+                     "Stream can't be null for '%s'",
+                     clazz.getSimpleName());
+        try {
+            return mapper.readValue(stream, clazz);
+        } catch (IOException e) {
+            throw new RuntimeException(String.format(
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/util/JsonUtil.java]: HugeException

File: hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java [hunk 8]
```diff
@@ -48,11 +49,14 @@
 
     private static volatile LicenseVerifier INSTANCE = null;
 
+    private static final Duration CHECK_INTERVAL = Duration.ofMinutes(10);
+    private static Instant lastCheckTime = Instant.now();
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java]: set volatile

File: hugegraph-api/src/main/java/com/baidu/hugegraph/config/ServerOptions.java [hunk 9]
```diff
@@ -48,6 +48,14 @@ public static synchronized ServerOptions instance() {
                     "http://127.0.0.1:8080"
             );
 
+    public static final ConfigOption<String> SERVER_ID =
+            new ConfigOption<>(
+                    "server.id",
+                    "The id of server.",
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/config/ServerOptions.java]: The id of hugegraph-server

File: hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+            }
+        }
+        return INSTANCE;
+    }
+
+    public static void verifyIfNeeded() {
+        Instant now = Instant.now();
+        Duration interval = Duration.between(lastCheckTime, now);
+        if (!interval.minus(CHECK_INTERVAL).isNegative()) {
+            LicenseVerifier.instance(null).verify();
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java]: may it cause exception at line 59 checkNotNull()?

File: hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    private final LicenseVerifyManager manager;
+
+    private LicenseVerifier(HugeConfig config) {
+        E.checkNotNull(config, "config");
+        this.verifyParam = buildVerifyParam(LICENSE_PARAM_PATH);
+        LicenseParam licenseParam = this.initLicenseParam(this.verifyParam);
+        this.manager = new LicenseVerifyManager(licenseParam, config);
+    }
+
+    public static LicenseVerifier instance(HugeConfig config) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java]: prefer add a method install(HugeConfig config), and remove config param from instance()

File: hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+import de.schlichtherle.license.DefaultCipherParam;
+import de.schlichtherle.license.DefaultKeyStoreParam;
+import de.schlichtherle.license.DefaultLicenseParam;
+import de.schlichtherle.license.KeyStoreParam;
+import de.schlichtherle.license.LicenseContent;
+import de.schlichtherle.license.LicenseParam;
+
+public class LicenseVerifier {
+
+    private static final Logger LOG = Log.logger(LicenseVerifier.class);
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java]: Log.logger(HugeGraph.class)

File: hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+    private static final String LICENSE_PARAM_PATH = "/verify-license.json";
+
+    private static volatile LicenseVerifier INSTANCE = null;
+
+    private static final Duration CHECK_INTERVAL = Duration.ofMinutes(10);
+    private static volatile Instant lastCheckTime = Instant.now();
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java]: remove static

File: hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifyManager.java [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    private void checkIp(ExtraParam param) throws LicenseContentException {
+        String expectIp = param.getIp();
+        if (StringUtils.isEmpty(expectIp)) {
+            return;
+        }
+        List<String> actualIps = this.machineInfo.getIpAddress();
+        if (!actualIps.contains(expectIp)) {
+            throw new LicenseContentException(String.format(
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifyManager.java]: add method newLicenseException(message, args)

File: hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java (hunk 10 above)
Comment (Linary) [reply hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java]: when pass param null, INSTANCE has been constructed

File: hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java (hunk 11 above)
Comment (Linary) [reply hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java]: put it in instance() can let HugeConfig as final field

File: hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java [hunk 15]
```diff
@@ -43,46 +44,46 @@
...
 
-    public static void verifyIfNeeded() {
+    public void verifyIfNeeded() {
         Instant now = Instant.now();
-        Duration interval = Duration.between(lastCheckTime, now);
+        Duration interval = Duration.between(this.lastCheckTime, now);
         if (!interval.minus(CHECK_INTERVAL).isNegative()) {
-            LicenseVerifier.instance(null).verify();
-            lastCheckTime = now;
+            LicenseVerifier.instance().verify();
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java]: this.verify();

File: hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifyManager.java [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+        this.checkIp(param);
+        this.checkMac(param);
+        this.checkCpu(param);
+        this.checkRam(param);
+        this.checkThreads(param);
+        this.checkMemory(param);
+    }
+
+    private String getServerId() {
+        return this.config.get(ServerOptions.SERVER_ID);
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifyManager.java]: add method this.config() and check not null (installed)

File: hugegraph-core/src/main/java/com/baidu/hugegraph/util/JsonUtil.java [hunk 17]
```diff
@@ -116,6 +118,18 @@ public static String toJson(Object object) {
         }
     }
 
+    public static <T> T  fromJson(InputStream stream, Class<T> clazz) {
+        E.checkState(stream != null,
+                     "Stream can't be null for '%s'",
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/util/JsonUtil.java]: Json stream

File: hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifyManager.java [hunk 18]
```diff
@@ -53,6 +56,11 @@ public void config(HugeConfig config) {
         this.config = config;
     }
 
+    public HugeConfig config() {
+        E.checkNotNull(this.config, "config");
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifyManager.java]: check state

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 19]
```diff
@@ -192,13 +199,12 @@ private void checkBackendVersionOrExit() {
...
-                LOG.error("The backend store of '{}' has not been initialized",
+                throw new BackendException(
+                          "The backend store of '%s' has not been initialized",
                           hugegraph.name());
-                System.exit(-1);
             }
             if (!info.checkVersion()) {
-                // Exit if versions are inconsistent
-                System.exit(-1);
+                throw new BackendException("Check backend store version failed");
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java]: The backend store version is inconsistent

File: hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+        }
+    }
+
+    public void verify() {
+        try {
+            LicenseContent content = this.manager.verify();
+            LOG.info("The license verification passed, valid for {} - {}",
+                     content.getNotBefore(), content.getNotAfter());
+        } catch (Exception e) {
+            throw new HugeException("The license verification failed!", e);
```
Comment (zhoney) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/license/LicenseVerifier.java]: prefer  "Failed to verify license"  to keep consistent

**Pull request merged**